### PR TITLE
apply default in objects and arrays

### PR DESCRIPTION
**Pull request by SiemelNaran**
At https://json-schema.org/understanding-json-schema/reference/generic.html they say of the default keyword

> The default keyword specifies a default value. This value is not used to fill in missing values during the validation process. Non-validation tools such as documentation generators or form generators may use this value to give hints to users about how to use a value. However, default is typically used to express that if a value is missing, then the value is semantically the same as if the value was present with the default value. The value of default should validate against the schema in which it resides, but that isn’t required.

So before this PR, the 'default' keyword is ignored during both validation and walking.  But this PR allow the user to apply defaults during walking.  They still have to set a flag indicating that they want defaults to be applied (and they can apply defaults only if property is missing, or if it missing or is explicitly null).

If JSON blobs were saved to the database, and the JSON schema changed to add defaults, the db migration tool could use read the JSON blobs and walk them against the schema again to apply defaults.

### CODE REVIEW

**Comment by SiemelNaran** [comment]
mvn test results in 4 errors

```
[ERROR] Failures: 
[ERROR]   OpenAPI30JsonSchemaTest.testDiscriminatorMapping:120->runTestFile:109 expected: <false> but was: <true>
[ERROR]   V4JsonSchemaTest.testNotValidator:155->BaseSuiteJsonSchemaTest.runTestFile:116 expected: <false> but was: <true>
[ERROR]   V7JsonSchemaTest.testNotValidator:272->BaseSuiteJsonSchemaTest.runTestFile:116 expected: <false> but was: <true>
[ERROR] Errors: 
[ERROR]   CollectorContextTest.testCollectorGetAll:109 » NullPointer
[INFO] 
[ERROR] Tests run: 310, Failures: 3, Errors: 1, Skipped: 46
```

File: src/main/java/com/networknt/schema/JsonSchema.java [hunk 1]
```diff
@@ -239,6 +239,20 @@ private boolean nodeContainsRef(String ref, JsonNode node) {
         return validators;
     }
 
+    /**
+     * A comparator that sorts validators, such such that 'properties' comes before 'required',
+     * so that we can apply default values before validating required.
+     */
+    private static Comparator<String> VALIDATOR_SORT = (lhs, rhs) -> {
```
Comment (SiemelNaran) [inline src/main/java/com/networknt/schema/JsonSchema.java:246]: This change also changes the order of validators, so the order of validation messages will change.  If anyone was expecting the old order, they will break. Fortunately mvn test passes, and my code uses Matchers.containsInAnyOrder.

Comment (stevehu) [reply src/main/java/com/networknt/schema/JsonSchema.java:246]: I don't think the order of the messages is important.

Comment (SiemelNaran) [reply src/main/java/com/networknt/schema/JsonSchema.java:246]: What about performance concerns?  Changing HashMap to TreeMap is changing O(1) to O(lg(N)).  Is there a performance test that can be run to test?

Comment (stevehu) [reply src/main/java/com/networknt/schema/JsonSchema.java:246]: The focus of this json-schema-validator is the small size of the JSON object and the number of the validators shouldn't that big. In most cases, we shouldn't worry about the performance. For the large-size JSON schemas, we might have an issue. I would suggest picking up one big JSON schema to test it with this [repo](https://github.com/networknt/json-schema-validator-perftest). If we see a performance issue, we can make it configurable to use either HashMap or TreeMap in the code. What do you think?

Comment (SiemelNaran) [reply src/main/java/com/networknt/schema/JsonSchema.java:246]: (1) Using json-schema-validator-perftest as is which is version 1.0.49 time is around 430ms

(2) Upgrading to version 1.0.63 by changing the version to
```
            <version>1.0.63</version>
```
and the time is now around 465ms.

(3) Using my code by adding

            <scope>system</scope>
            <systemPath>/Users/snaran/main/Software/json-schema-validator/target/json-schema-validator-1.0.63.jar</systemPath>

now time is around 430ms.  Maybe the TreeMap makes things faster.  Hash map has to hash the whole string, which is O(N) of the whole string, whereas TreeMap only has to compare chars till it finds the first difference, so in theory TreeMap could be faster sometimes.  But who knows, just guessing.

**Comment by SiemelNaran** [comment]
Reminder about this review in case you missed it @prashanthjos .  @stevehu I did the performance tests and replied in the thread, and to my surprise found that the code is faster after my changes.

**Comment by stevehu** [comment]
@SiemelNaran That is good news :) Let us know when you think the PR is completed so that we can start the review. Thanks a lot for the effort.

**Comment by SiemelNaran** [comment]
> @SiemelNaran That is good news :) Let us know when you think the PR is completed so that we can start the review. Thanks a lot for the effort.

It is ready.

**Pull request merged**